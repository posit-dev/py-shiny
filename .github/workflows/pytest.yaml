name: Run tests

on:
  workflow_dispatch:
  push:
    branches: ["main", "rc-*"]
  pull_request:
    types: [opened, synchronize, reopened, ready_for_review]
  release:
    types: [published]
  schedule:
    - cron: "0 8 * * *"

jobs:
  check:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        # "3.10" must be a string; otherwise it is interpreted as 3.1.
        python-version: ["3.12", "3.11", "3.10", "3.9", "3.8"]
        os: [ubuntu-latest, windows-latest, macOS-latest]
        exclude:
          - python-version: ${{ github.event.pull_request.draft && '3.11' }}
          - python-version: ${{ github.event.pull_request.draft && '3.10' }}
          - python-version: ${{ github.event.pull_request.draft && '3.9' }}
          - os: ${{ github.event.pull_request.draft && 'windows-latest' }}
          - os: ${{ github.event.pull_request.draft && 'macOS-latest' }}

      fail-fast: false

    steps:
      - uses: actions/checkout@v4
      - name: Setup py-shiny
        id: install
        uses: ./.github/py-shiny/setup
        with:
          python-version: ${{ matrix.python-version }}

      - name: Run unit tests
        if: steps.install.outcome == 'success' && (success() || failure())
        run: |
          make check-tests

      - name: Type check
        if: steps.install.outcome == 'success' && (success() || failure())
        run: |
          make check-types

      - name: Lint code
        if: steps.install.outcome == 'success' && (success() || failure())
        run: |
          make check-lint

      - name: Verify code formatting
        if: steps.install.outcome == 'success' && (success() || failure())
        run: |
          make check-format

  pypi:
    name: "Deploy to PyPI"
    runs-on: ubuntu-latest
    if: github.event_name == 'release'
    needs: [check]
    steps:
      - uses: actions/checkout@v4
      - name: "Set up Python 3.10"
        uses: actions/setup-python@v4
        with:
          python-version: "3.10"
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install https://github.com/rstudio/py-htmltools/tarball/main
          make install-deps
          make install
      - name: "Build Package"
        run: |
          make dist

      # test deploy ----
      - name: "Test Deploy to PyPI"
        uses: pypa/gh-action-pypi-publish@release/v1
        if: startsWith(github.event.release.name, 'TEST')
        with:
          user: __token__
          password: ${{ secrets.PYPI_TEST_API_TOKEN }}
          repository-url: https://test.pypi.org/legacy/

      ## prod deploy ----
      - name: "Deploy to PyPI"
        uses: pypa/gh-action-pypi-publish@release/v1
        if: startsWith(github.event.release.name, 'shiny')
        with:
          user: __token__
          password: ${{ secrets.PYPI_API_TOKEN }}

  playwright-shiny:
    if: github.event_name != 'release'
    runs-on: ubuntu-latest
    strategy:
      matrix:
        python-version: ["3.12", "3.11", "3.10", "3.9", "3.8"]
        browser: ["chromium", "firefox", "webkit"]
        exclude:
          - python-version: ${{ github.event.pull_request.draft && '3.11' }}
          - python-version: ${{ github.event.pull_request.draft && '3.10' }}
          - python-version: ${{ github.event.pull_request.draft && '3.9' }}
          - browser: ${{ github.event.pull_request.draft && 'firefox' }}
          - browser: ${{ github.event.pull_request.draft && 'webkit' }}
          # There are many unexplained tests that fail on webkit w/ python 3.8, 3.9
          # Given the more recent versions of python work, we will exclude this combination
          - browser: "webkit"
            python-version: "3.8"
          - browser: "webkit"
            python-version: "3.9"
      fail-fast: false

    steps:
      - uses: actions/checkout@v4
      - name: Setup py-shiny
        uses: ./.github/py-shiny/setup
        with:
          python-version: ${{ matrix.python-version }}

      - name: Determine browsers for testing
        uses: ./.github/py-shiny/pytest-browsers
        id: browsers
        with:
          browser: ${{ matrix.browser }}
          # If anything other than `true`, it will heavily reduce webkit performance
          # Related: https://github.com/microsoft/playwright/issues/18119
          disable-playwright-diagnostics: ${{ matrix.browser == 'webkit' || matrix.browser == 'firefox' }}

      - name: Run End-to-End tests
        timeout-minutes: 60
        run: |
          make playwright-shiny SUB_FILE=". --numprocesses 3 ${{ steps.browsers.outputs.playwright-diagnostic-args }}" ${{ steps.browsers.outputs.browsers }}
      - uses: actions/upload-artifact@v4
        if: failure() && steps.browsers.outputs.has-playwright-diagnostics
        with:
          name: "playright-shiny-${{ runner.os }}-${{ matrix.python-version }}-${{ matrix.browser }}-results"
          path: test-results/
          retention-days: 5

  playwright-examples:
    if: github.event_name != 'release'
    runs-on: ubuntu-20.04
    strategy:
      matrix:
        python-version: ["3.12", "3.11", "3.10", "3.9", "3.8"]
        browser: ["chromium", "firefox", "webkit"]
        exclude:
          - python-version: ${{ github.event.pull_request.draft && '3.11' }}
          - python-version: ${{ github.event.pull_request.draft && '3.10' }}
          - python-version: ${{ github.event.pull_request.draft && '3.9' }}
          - browser: ${{ github.event.pull_request.draft && 'firefox' }}
          - browser: ${{ github.event.pull_request.draft && 'webkit' }}
      fail-fast: false

    steps:
      - uses: actions/checkout@v4
      - name: Setup py-shiny
        uses: ./.github/py-shiny/setup
        with:
          python-version: ${{ matrix.python-version }}
      - name: Determine browsers for testing
        uses: ./.github/py-shiny/pytest-browsers
        id: browsers
        with:
          browser: ${{ matrix.browser }}
          # If anything other than `true`, it will heavily reduce webkit performance
          # Related: https://github.com/microsoft/playwright/issues/18119
          disable-playwright-diagnostics: ${{ matrix.browser == 'webkit' || matrix.browser == 'firefox' }}

      - name: Install node.js
        uses: actions/setup-node@v4
        with:
          node-version: "18"
          cache: npm
          cache-dependency-path: examples/brownian/shinymediapipe/package-lock.json
      - name: Install node.js package
        working-directory: examples/brownian/shinymediapipe
        run: |
          npm ci

      - name: Run example app tests
        timeout-minutes: 60
        run: |
          make playwright-examples SUB_FILE=". --numprocesses 3 ${{ steps.browsers.outputs.playwright-diagnostic-args }}" ${{ steps.browsers.outputs.browsers }}
      - uses: actions/upload-artifact@v4
        if: failure() && steps.browsers.outputs.has-playwright-diagnostics
        with:
          name: "playright-examples-${{ runner.os }}-${{ matrix.python-version }}-${{ matrix.browser }}-results"
          path: test-results/
          retention-days: 5

  playwright-deploys-precheck:
    if: github.event_name != 'release'
    runs-on: ubuntu-latest
    strategy:
      matrix:
        # Matches deploy server python version
        python-version: ["3.10"]
      fail-fast: false

    steps:
      - uses: actions/checkout@v4
      - name: Setup py-shiny
        uses: ./.github/py-shiny/setup
        with:
          python-version: ${{ matrix.python-version }}
      - name: Determine browsers for testing
        uses: ./.github/py-shiny/pytest-browsers
        id: browsers
        with:
          all-browsers: ${{ ! github.event.pull_request.draft }}

      - name: Test that deployable example apps work
        timeout-minutes: 5 # ~10s locally
        env:
          DEPLOY_APPS: "false"
        run: |
          make playwright-deploys

      - uses: actions/upload-artifact@v4
        if: failure()
        with:
          name: "playright-examples-${{ runner.os }}-${{ matrix.python-version }}-results"
          path: test-results/
          retention-days: 5
