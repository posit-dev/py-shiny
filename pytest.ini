[pytest]
asyncio_mode=strict
