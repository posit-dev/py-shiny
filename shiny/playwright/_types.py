"""Facade classes for working with Shiny inputs/outputs in Playwright"""

from __future__ import annotations

import typing
from typing import Optional

OptionalStr = Optional[str]
OptionalInt = Optional[int]
OptionalFloat = Optional[float]
OptionalBool = Optional[bool]

PatternStr = typing.Pattern[str]
PatternOrStr = typing.Union[str, PatternStr]
ListPatternOrStr = typing.Union[
    typing.List[PatternOrStr], typing.List[str], typing.List[PatternStr]
]
AttrValue = typing.Union[PatternOrStr, None]
StyleValue = typing.Union[PatternOrStr, None]

Timeout = typing.Union[float, None]
