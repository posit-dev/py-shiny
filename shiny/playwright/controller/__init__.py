from ._input_buttons import (
    InputActionButton,
    InputActionLink,
    InputDarkMode,
    InputFile,
    InputTaskButton,
)

from ._input_fields import (
    InputDate,
    InputDateRange,
    InputNumeric,
    InputPassword,
    InputText,
    InputTextArea,
)

from ._input_controls import (
    InputCheckbox,
    InputCheckboxGroup,
    InputRadioButtons,
    InputSelect,
    InputSelectize,
    InputSlider,
    InputSliderRange,
    InputSwitch,
)

from ._overlay import (
    Popover,
    Tooltip,
)

from ._layout import (
    Sidebar,
)

from ._accordion import (
    Accordion,
    AccordionPanel,
)

from ._card import Card, ValueBox

from ._file import (
    DownloadButton,
    DownloadLink,
)

from ._chat import (
    Chat,
)
from ._navs import (
    NavPanel,
    NavsetBar,
    NavsetCardPill,
    NavsetCardTab,
    NavsetCardUnderline,
    NavsetHidden,
    NavsetPill,
    NavsetPillList,
    NavsetTab,
    NavsetUnderline,
)
from ._output import (
    OutputCode,
    OutputDataFrame,
    OutputImage,
    OutputPlot,
    OutputTable,
    OutputText,
    OutputTextVerbatim,
    OutputUi,
)

__all__ = [
    "InputActionButton",
    "InputActionLink",
    "InputCheckbox",
    "InputCheckboxGroup",
    "InputDarkMode",
    "InputDate",
    "InputDateRange",
    "InputFile",
    "InputNumeric",
    "InputPassword",
    "InputRadioButtons",
    "InputSelect",
    "InputSelectize",
    "InputSlider",
    "InputSliderRange",
    "InputSwitch",
    "InputTaskButton",
    "InputText",
    "InputTextArea",
    "OutputCode",
    "OutputDataFrame",
    "OutputImage",
    "OutputPlot",
    "OutputTable",
    "OutputText",
    "OutputTextVerbatim",
    "OutputUi",
    "ValueBox",
    "Card",
    "Chat",
    "Accordion",
    "AccordionPanel",
    "Sidebar",
    "Popover",
    "Tooltip",
    "NavPanel",
    "NavsetBar",
    "NavsetCardPill",
    "NavsetCardTab",
    "NavsetCardUnderline",
    "NavsetHidden",
    "NavsetPill",
    "NavsetPillList",
    "NavsetTab",
    "NavsetUnderline",
    "DownloadButton",
    "DownloadLink",
]
