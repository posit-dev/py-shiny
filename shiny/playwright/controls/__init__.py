from ._controls import (
    InputActionButton,
    InputActionLink,
    InputCheckbox,
    InputCheckboxGroup,
    InputDarkMode,
    InputDate,
    InputDateRange,
    InputFile,
    InputNumeric,
    InputPassword,
    InputRadioButtons,
    InputSelect,
    InputSelectize,
    InputSlider,
    InputSliderRange,
    InputSwitch,
    InputTaskButton,
    InputText,
    InputTextArea,
    OutputCode,
    OutputDataFrame,
    OutputImage,
    OutputPlot,
    OutputTable,
    OutputText,
    OutputTextVerbatim,
    OutputUi,
    ValueBox,
    Card,
    Accordion,
    Sidebar,
    Popover,
    Tooltip,
    LayoutNavItem,
    LayoutNavSetBar,
    LayoutNavSetCardPill,
    LayoutNavSetCardTab,
    LayoutNavSetCardUnderline,
    LayoutNavSetHidden,
    LayoutNavSetPill,
    LayoutNavSetPillList,
    LayoutNavsetTab,
    LayoutNavSetUnderline,
    DownloadButton,
    DownloadLink,
)

__all__ = [
    "InputActionButton",
    "InputActionLink",
    "InputCheckbox",
    "InputCheckboxGroup",
    "InputDarkMode",
    "InputDate",
    "InputDateRange",
    "InputFile",
    "InputNumeric",
    "InputPassword",
    "InputRadioButtons",
    "InputSelect",
    "InputSelectize",
    "InputSlider",
    "InputSliderRange",
    "InputSwitch",
    "InputTaskButton",
    "InputText",
    "InputTextArea",
    "OutputCode",
    "OutputDataFrame",
    "OutputImage",
    "OutputPlot",
    "OutputTable",
    "OutputText",
    "OutputTextVerbatim",
    "OutputUi",
    "ValueBox",
    "Card",
    "Accordion",
    "Sidebar",
    "Popover",
    "Tooltip",
    "LayoutNavItem",
    "LayoutNavSetBar",
    "LayoutNavSetCardPill",
    "LayoutNavSetCardTab",
    "LayoutNavSetCardUnderline",
    "LayoutNavSetHidden",
    "LayoutNavSetPill",
    "LayoutNavSetPillList",
    "LayoutNavsetTab",
    "LayoutNavSetUnderline",
    "DownloadButton",
    "DownloadLink",
]
