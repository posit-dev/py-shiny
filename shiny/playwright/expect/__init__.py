from ._expect import (
    expect_not_to_have_attribute,
    expect_not_to_have_class,
    expect_to_have_class,
    expect_to_have_style,
    expect_not_to_have_style,
)
from ._expect_to_change import expect_to_change

__all__ = [
    "expect_to_change",
    "expect_not_to_have_attribute",
    "expect_to_have_class",
    "expect_not_to_have_class",
    "expect_to_have_style",
    "expect_not_to_have_style",
]
