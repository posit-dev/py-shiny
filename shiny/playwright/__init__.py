try:
    import playwright  # noqa: F401 # pyright: ignore[reportUnusedImport, reportMissingTypeStubs]
except ImportError:
    raise ImportError(
        "The shiny.playwright module requires the playwright package to be installed."
        " Please install it with this command:"
        "\n\n    pip install playwright"
    )
# If `pytest` is installed...
try:
    import pytest  # noqa: F401 # pyright: ignore[reportUnusedImport, reportMissingTypeStubs]

    # At this point, `playwright` and `pytest` are installed.
    # Try to make sure `pytest-playwright` is installed
    try:
        import pytest_playwright  # noqa: F401 # pyright: ignore[reportUnusedImport, reportMissingTypeStubs]

    except ImportError:
        raise ImportError(
            "If you are using pytest to test your app,"
            " you can install the pytest-playwright shim package with this command:",
            "\n\n    pip install pytest-playwright",
        )
except ImportError:
    pass
from . import controller, expect

__all__ = ["expect", "controller"]
