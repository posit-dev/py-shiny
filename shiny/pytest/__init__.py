try:
    import pytest  # noqa: F401 # pyright: ignore[reportUnusedImport, reportMissingTypeStubs]
except ImportError:
    raise ImportError(
        "The shiny.pytest module requires the pytest package to be installed."
        " Please install it with this command:"
        "\n\n    pip install pytest"
    )

from ._fixture import ScopeName, create_app_fixture

__all__ = ("create_app_fixture", "ScopeName")
