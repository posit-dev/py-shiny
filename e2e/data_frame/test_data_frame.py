# pyright: reportUnknownMemberType=false


import re
import time
from typing import Any, Callable

import pytest
from conftest import ShinyAppProc, create_example_fixture, expect_to_change
from controls import InputSelectize, InputSwitch
from playwright.sync_api import Locator, Page, expect

data_frame_app = create_example_fixture("dataframe")


@pytest.fixture
def grid(page: Page) -> Locator:
    return page.locator("#grid")


@pytest.fixture
def grid_container(page: Page, grid: Locator) -> Locator:
    return grid.locator("> div > div.shiny-data-grid")


@pytest.fixture
def summary(page: Page, grid: Locator) -> Locator:
    return grid.locator("div.shiny-data-grid-summary")


@pytest.fixture
def scroll_to_end(page: Page, grid_container: Locator) -> Callable[[], None]:
    def do():
        grid_container.locator("tbody tr:first-child td:first-child").click()
        page.keyboard.press("End")

    return do


@pytest.mark.flaky
def test_grid_mode(
    page: Page, data_frame_app: ShinyAppProc, grid: Locator, grid_container: Locator
):
    page.goto(data_frame_app.url)

    # Ensure the output was found by Shiny
    expect(grid).to_have_class(re.compile(r"\bshiny-bound-output\b"))

    # Expect the data frame to be in grid mode by default
    expect(grid_container).not_to_have_class(re.compile(r"\bshiny-data-grid-table\b"))
    expect(grid_container).to_have_class(re.compile(r"\bshiny-data-grid-grid\b"))


@pytest.mark.flaky
def test_summary_navigation(
    page: Page, data_frame_app: ShinyAppProc, grid_container: Locator, summary: Locator
):
    page.goto(data_frame_app.url)

    # Check that summary responds to navigation
    expect(summary).to_have_text("Viewing rows 1 through 10 of 20")
    # Put focus in the table and hit End keystroke
    grid_container.locator("tbody tr:first-child td:first-child").click()
    with expect_to_change(lambda: summary.inner_text()):
        page.keyboard.press("End")
    # Ensure that summary updated
    expect(summary).to_have_text("Viewing rows 11 through 20 of 20")


@pytest.mark.flaky
def test_full_width(page: Page, data_frame_app: ShinyAppProc, grid_container: Locator):
    page.goto(data_frame_app.url)

    def get_width() -> float:
        rect = grid_container.bounding_box()
        return rect.get("width", 0) if rect else 0

    width1 = get_width()
    # Switch to narrow mode
    with expect_to_change(get_width):
        InputSwitch(page, "fullwidth").toggle()
    width2 = get_width()

    assert width2 < width1

    # Switch back to full width
    InputSwitch(page, "fullwidth").toggle()


@pytest.mark.flaky
def test_table_switch(
    page: Page,
    data_frame_app: ShinyAppProc,
    grid: Locator,
    grid_container: Locator,
    summary: Locator,
    scroll_to_end: Callable[[], None],
):
    page.goto(data_frame_app.url)
    select_dataset = InputSelectize(page, "dataset")

    scroll_to_end()

    # Switch to table
    InputSwitch(page, "gridstyle").toggle()
    expect(grid_container).not_to_have_class(re.compile(r"\bshiny-data-grid-grid\b"))
    expect(grid_container).to_have_class(re.compile(r"\bshiny-data-grid-table\b"))

    # Switching modes resets scroll
    expect(summary).to_have_text("Viewing rows 1 through 10 of 20")

    scroll_to_end()
    expect(summary).to_have_text("Viewing rows 11 through 20 of 20")

    # Switch datasets to much longer one
    select_dataset.set("diamonds")
    expect(summary).to_have_text("Viewing rows 1 through 10 of 53940")


@pytest.mark.flaky
def test_sort(
    page: Page,
    data_frame_app: ShinyAppProc,
    grid_container: Locator,
):
    page.goto(data_frame_app.url)
    select_dataset = InputSelectize(page, "dataset")
    select_dataset.set("diamonds")

    # Test sorting
    header_clarity = grid_container.locator("tr:first-child th:nth-child(4)")
    first_cell_clarity = grid_container.locator("tr:first-child td:nth-child(4)")
    expect(first_cell_clarity).to_have_text("SI2")
    header_clarity.click()
    expect(first_cell_clarity).to_have_text("I1")
    header_clarity.click()
    expect(first_cell_clarity).to_have_text("VVS2")

    # Test multi-sorting
    depth_clarity = grid_container.locator("tr:first-child th:nth-child(5)")
    first_cell_depth = grid_container.locator("tr:first-child td:nth-child(5)")
    depth_clarity.click(modifiers=["Shift"])
    expect(first_cell_depth).to_have_text("67.6")


@pytest.mark.flaky
def test_multi_selection(
    page: Page, data_frame_app: ShinyAppProc, grid_container: Locator, snapshot: Any
):
    page.goto(data_frame_app.url)
    first_cell = grid_container.locator("tbody tr:first-child td:first-child")

    def detail_text():
        return page.locator("#detail").inner_text()

    expect(first_cell).to_have_text("1")
    with expect_to_change(detail_text):
        first_cell.click()

    assert detail_text() == snapshot

    kb = page.keyboard
    with expect_to_change(lambda: grid_container.locator(":focus").inner_text()):
        kb.press("ArrowDown")
    with expect_to_change(lambda: grid_container.locator(":focus").inner_text()):
        kb.press("ArrowDown")
    with expect_to_change(detail_text):
        kb.press("Space")
    with expect_to_change(lambda: grid_container.locator(":focus").inner_text()):
        kb.press("ArrowDown")
    with expect_to_change(detail_text):
        kb.press("Enter")
    assert detail_text() == snapshot


@pytest.mark.flaky
def test_single_selection(
    page: Page, data_frame_app: ShinyAppProc, grid_container: Locator, snapshot: Any
):
    page.goto(data_frame_app.url)
    InputSelectize(page, "selection_mode").set("single")
    first_cell = grid_container.locator("tbody tr:first-child td:first-child")

    def detail_text():
        return page.locator("#detail").inner_text()

    expect(first_cell).to_have_text("1")
    with expect_to_change(detail_text):
        first_cell.click()

    assert detail_text() == snapshot

    kb = page.keyboard
    with expect_to_change(detail_text):
        kb.press("ArrowDown")
    with expect_to_change(detail_text):
        kb.press("ArrowDown")
    with expect_to_change(detail_text):
        kb.press("Space")
    with expect_to_change(lambda: grid_container.locator(":focus").inner_text()):
        kb.press("ArrowDown")
    with expect_to_change(detail_text):
        kb.press("Enter")
    assert detail_text() == snapshot


def retry_with_timeout(timeout: float = 30):
    def decorator(func: Callable[[], None]) -> None:
        def exec() -> None:
            start = time.time()
            while True:
                try:
                    return func()
                except AssertionError as e:
                    if time.time() - start > timeout:
                        raise e
                    time.sleep(0.1)

        exec()

    return decorator
