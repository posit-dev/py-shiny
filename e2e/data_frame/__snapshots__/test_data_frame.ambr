# serializer version: 1
# name: test_multi_selection[chromium]
  '''
     subidr    attnr  num1  num2  num3
  0       1  divided     2   4.0     7
  '''
# ---
# name: test_multi_selection[chromium].1
  '''
     subidr    attnr  num1  num2  num3
  0       1  divided     2   4.0     7
  2       3  divided     3   5.0     6
  3       4  divided     5   7.0     5
  '''
# ---
# name: test_multi_selection[firefox]
  '''
     subidr    attnr  num1  num2  num3
  0       1  divided     2   4.0     7
  '''
# ---
# name: test_multi_selection[firefox].1
  '''
     subidr    attnr  num1  num2  num3
  0       1  divided     2   4.0     7
  2       3  divided     3   5.0     6
  3       4  divided     5   7.0     5
  '''
# ---
# name: test_multi_selection[webkit]
  '''
     subidr    attnr  num1  num2  num3
  0       1  divided     2   4.0     7
  '''
# ---
# name: test_multi_selection[webkit].1
  '''
     subidr    attnr  num1  num2  num3
  0       1  divided     2   4.0     7
  2       3  divided     3   5.0     6
  3       4  divided     5   7.0     5
  '''
# ---
# name: test_single_selection[chromium]
  '''
     subidr    attnr  num1  num2  num3
  0       1  divided     2   4.0     7
  '''
# ---
# name: test_single_selection[chromium].1
  '''
     subidr    attnr  num1  num2  num3
  3       4  divided     5   7.0     5
  '''
# ---
# name: test_single_selection[firefox]
  '''
     subidr    attnr  num1  num2  num3
  0       1  divided     2   4.0     7
  '''
# ---
# name: test_single_selection[firefox].1
  '''
     subidr    attnr  num1  num2  num3
  3       4  divided     5   7.0     5
  '''
# ---
# name: test_single_selection[webkit]
  '''
     subidr    attnr  num1  num2  num3
  0       1  divided     2   4.0     7
  '''
# ---
# name: test_single_selection[webkit].1
  '''
     subidr    attnr  num1  num2  num3
  3       4  divided     5   7.0     5
  '''
# ---
